Fix strtr inferences

VincentLanglet opened the pull request:
Closes https://github.com/phpstan/phpstan/issues/13708

VincentLanglet committed:
Fix strtr inferences

staabm commented on tests/PHPStan/Rules/Methods/data/bug-13708.php:
```diff
@@ -0,0 +1,23 @@
+	public function doStuff(): void
+	{
+		$this->takeNonEmpty(
+			strtr('change {me}', ['{me}' => 'me'])
```
Would be good to have an assertType() somewhere so we see the concrete type when all args are constant like here

VincentLanglet replied:
I have added multiple assertion in the existing nsrt file https://github.com/phpstan/phpstan-src/pull/4466/commits/93b42a2e34eaa04095d1d7f1cf8888a03724571f

VincentLanglet committed:
Add assertion

ondrejmirtes commented:
Escaped mutants here. This means tests need to be written so that when the code changes like that, there will be some tests that will fail. Possibly the code logic might be wrong too.

VincentLanglet commented:
> Escaped mutants here. This means tests need to be written so that when the code changes like that, there will be some tests that will fail. Possibly the code logic might be wrong too.

I'm not sure what should be expected here.

The mutant can only occurs if we pass `string|array` as second param and wether it infers non-empty-string or not seems kinda useless to me because it's already reported as an error by PHPStan which should be fixed first https://phpstan.org/r/1b1f1f8a-dbd4-417c-b251-be553d4d64d3.

Also if I try to support "maybe array" with something like
```
TypeCombinator::intersect(new ArrayType(new MixedType(), new MixedType()), $secondArgumentType)
```
I'll still get "false positive" escaped mutant because now `$type->isArray()->maybe()` will be useless and therefor `->yes()` or `! -> no()` are equivalent.

ondrejmirtes commented:
Maybe array is also for example `mixed` or `array|null`.

> kinda useless to me because it's already reported as an error by PHPStan

Type inference and error reporting are two independent different parts of PHPStan. Type inference should still do something sensible even if the user passes a wrong type there. Especially when the wrong type might not even be reported on a lower rule level.

VincentLanglet committed:
Improve maybe inferences

VincentLanglet committed:
Fix cs

VincentLanglet committed:
Try

VincentLanglet commented:
No mutant escaped now @ondrejmirtes :) 

The lint 8.0 failure seems unrelated/random

VincentLanglet commented:
> No mutant escaped now @ondrejmirtes :)
> 
> The lint 8.0 failure seems unrelated/random

I also rebased and fixed all my existing PR if you have time for a look https://github.com/phpstan/phpstan-src/pulls/VincentLanglet.
I commented those with failing build, some failure are unrelated/random ; and some mutant are (I think) not fixable.

staabm commented on tests/PHPStan/Analyser/nsrt/strtr.php:
```diff
@@ -24,4 +25,27 @@ function doFoo(string $s, $nonEmptyString, $nonFalseyString) {
+
+	assertType('non-empty-string', strtr($nonFalseyString, [$s => $nonEmptyString]));
+	assertType('non-falsy-string', strtr($nonFalseyString, [$nonEmptyString => $nonFalseyString]));
+	assertType('non-falsy-string', strtr($nonFalseyString, [$nonFalseyString => $nonFalseyString]));
```
```suggestion
	assertType('non-falsy-string', strtr($nonFalseyString, [$nonFalseyString => $nonFalseyString]));

	// see https://3v4l.org/jCUhW/rfc#vgit.master
	assertType('string', strtr($nonFalseyString, [$nonFalseyString => $s]));
	assertType('string', strtr($nonEmptyString, [$nonEmptyString => $s]));
```

VincentLanglet replied:
Done

VincentLanglet committed:
Add tests
